enhancements/kube-apiserver/stability-using-dead-api.md

```diff
@@ -0,0 +1,213 @@
+---
+title: stability-who-used-the-dying-api
+authors:
+  - "@deads2k"
+reviewers:
+  - TBD
+  - "@alicedoe"
+approvers:
+  - TBD
+  - "@oscardoe"
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also:
+  - "/enhancements/this-other-neat-thing.md"  
+replaces:
+  - "/enhancements/that-less-than-great-idea.md"
+superseded-by:
+  - "/enhancements/our-past-effort.md"
+---
+
+# Stability: Who Used the Dying API
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding
+to implement the design.  For instance, 
+ > 1. This requires exposing previously private resources which contain sensitive
+  information.  Can we do this? 
+
+## Summary
+
+Provide a way for customers to know
+1. they are using a deprecated API
+2. which deprecated API are they using
+3. which user is using which API
+
+## Motivation
+
+Kube has started removing old APIs, users need to know what will fail before they upgrade.
+
+
+### Goals
+
+1. Create an automated way to generate alerts when updating kube levels.
+2. Provide a way to determine which users are using which deprecated APIs.
+3. Inform a decision about whether or not to allow upgrades.
+
+### Non-Goals
+
+1. Prevent deprecated APIs from being removed.
+2. This enhancement doesn't detail addition features we could add to upstream via a discovery doc and/or CRDs to make
+   deprecation more obvious and discoverable at runtime.
+   It simply works with what we have a available today.
+
+## Proposal
+
+1. Write a generator for alerts in kas-o (or openshift/kubernetes), that loads the current scheme (we functions for
+   this in openshift/api), navigates each type, determine which are deprecated and which are about to be removed.
+   This information can then be used to generate specific alerts and/or metrics.
+2. In the kas-o, if a "deprecated API used" alert has been fired, we have sufficient information to parse an
+   audit log and summarize the top X users of deprecated APIs.
```

Comment:
parsing audit log sounds pretty hacky to me, isn't there a better way? (admission/etc).

Reply:
> parsing audit log sounds pretty hacky to me, isn't there a better way? (admission/etc).

Admission doesn't cover get/list/watch.  I don't know of a better way to do it.  Perhaps a different audit sink, but that seems more fragile overall.